### PR TITLE
James 2286 define a API for mailrepository and provide contract tests

### PR DESCRIPTION
**Pull request by mbaechler**
This is a first (probably mergeable) step that brings junit5 and interface testing.

### CODE REVIEW

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        return InputStreams.create(headers + headerBodySeparator + body + end, StandardCharsets.UTF_8);
+    }
+
+    default void checkMailEquality(Mail actual, Mail expected) {
+        assertAll(
+            () -> assertThat(actual.getMessage().getContent()).isEqualTo(expected.getMessage().getContent()),
+            () -> assertThat(actual.getMessageSize()).isEqualTo(expected.getMessageSize()),
+            () -> assertThat(actual.getName()).isEqualTo(expected.getName()),
+            () -> assertThat(actual.getState()).isEqualTo(expected.getState()),
+            () -> assertThat(actual.getAttribute(TEST_ATTRIBUTE)).isEqualTo(expected.getAttribute(TEST_ATTRIBUTE))
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: You miss many fields: lastUpdated, errormessage,  remoteAddr, remoteHost, etc...

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    @Test
+    default void listShouldReturnStoredMailsKeys() throws Exception {
+        MailRepository testee = retrieveRepository();
+        testee.store(createMail("mail1"));
+        testee.store(createMail("mail2"));
+        assertThat(testee.list()).containsExactly("mail1", "mail2");
+    }
+
+    @Test
+    @Disabled("JDBC doesn't update the message Content")
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: Shouldn't this only be disabled for JDBC ?

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
+package org.apache.james.mailrepository;
```
Comment (aduprat) [inline server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: license missing

Comment (chibenwa) [inline server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: Missing license. You should solve this problem!

Comment (chibenwa) [reply server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: Maybe a checkstyle?

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+import org.apache.james.mailrepository.api.MailRepository;
+import org.apache.james.mime4j.io.InputStreams;
+import org.apache.james.server.core.MailImpl;
+import org.apache.mailet.Mail;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+
+import com.google.common.collect.ImmutableList;
+
+public interface MailRepositoryTest {
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: Is it a test? Can I run it?

I like the standard Junit denomination: "**MailRepositoryContract**". Then implementation can provide test classes that implements these contracts, like **JDBCMailRepositoryTest**

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java (hunk 1 above)
Comment (mbaechler) [reply server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: No I don't think so. the commit is "port existing tests" and the existing test does exactly these assertions. I'll do more when I'll be covering MailRepository more seriously.

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java (hunk 4 above)
Comment (mbaechler) [reply server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: As discussed during meeting, I'll go for Contract

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java (hunk 2 above)
Comment (mbaechler) [reply server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: I'll remove this, it makes no sense as it doesn't even run on jdbc at this point of the history

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java (hunk 3 above)
Comment (mbaechler) [reply server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: Yes, that's something we can do with checkstyle (http://checkstyle.sourceforge.net/config_header.html#Header). Do you want to handle such a PR ?

**Comment by mbaechler** [comment]
pushed -f because class renaming is not very fixup friendly

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java (hunk 2 above)
Comment (mbaechler) [reply server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: changed

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java (hunk 3 above)
Comment (mbaechler) [reply server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: fixed

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java (hunk 4 above)
Comment (mbaechler) [reply server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: changed

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        return mail;
+    }
+
+    default InputStream generateMailContent() {
+        String headers = new StringJoiner("\r\n")
+            .add("Subject: test")
+            .add("Content-Type: text/plain")
+            .toString();
+        String headerBodySeparator = "\r\n\r\n";
+        String body = "original body" + RandomStringUtils.randomAlphabetic(10);
```
Comment (rouazana) [inline server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: why introducing randomness?
I guess as this is just porting a test you miss a remove to clearly show that you are just moving some tests, no?

Comment (mbaechler) [reply server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: To be honest, I've read "Found more than one message after storing 2 times the SAME message: it MUST update the previous" in the Abstract version, but the assertion was not testing that. So I wrote a test about that to understand what it probably expected to test back then.
The problem is : you do make mails with different contents, you write a new method that takes a body ? I don't like that very much, method overloading is painful. Modify Mail then call store with the modified version ? I don't trust Mail implementation enough for that. Write a Mail.Builder ? I hesitate for some time but Hell, all these classes are a mess, I'll take forever ):

So sadly, I've chose the quickest solution for now. I'm pretty sure I'll have to challenge that before the end of the ticket.

Do you think we can merge that for now ?

Comment (chibenwa) [reply server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: How about making FakeMail serializable; and relying on FakeMail? Some other fixes along the way, but I succeeded writing all MailQueue Tests with FakeMail.

Comment (mbaechler) [reply server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: It's one more point of frustration : Mail is just a datastructure, it doesn't make sense to have an interface and several impl.
I digged that yesterday but I'd rather merge FakeMail and MailImpl than use FakeMail more. I'm really not confidence use FakeMail or MailImpl in tests gives the same result today

**Comment by aduprat** [comment]
JUnit 4 & 5 tests are running with the last commit (9722dbd), checked with:
```
✗ mvn clean install -Drat.numUnapprovedLicenses=100 -am -pl server/data/data-jdbc
```

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java (hunk 5 above)
Comment (rouazana) [reply server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java]: let's go just adding a method that take a content in parameter to build different mails, I find it cleaner than this randomness

**Comment by chibenwa** [comment]
New one ^^

```
[2ae1361e287a4e847746b1ddb23834c32f1c2cb1] Tests in error: 
[2ae1361e287a4e847746b1ddb23834c32f1c2cb1]   CassandraSharingTest » IllegalState A scoping block is already in progress
[2ae1361e287a4e847746b1ddb23834c32f1c2cb1]   CassandraSharingTest » OutOfMemory Java heap space
[2ae1361e287a4e847746b1ddb23834c32f1c2cb1]   CassandraSharingTest » IllegalState A scoping block is already in progress
```

test this please

File: pom.xml [hunk 6]
```diff
@@ -2541,13 +2567,30 @@
...
                         <argLine>-Xms512m -Xmx1024m</argLine>
                         <reuseForks>false</reuseForks>
                         <!-- Fail tests longer than 2 hours, prevent form random locking tests -->
                         <forkedProcessTimeoutInSeconds>7200</forkedProcessTimeoutInSeconds>
                     </configuration>
+                    <dependencies>
+                        <dependency>
+                            <groupId>org.junit.jupiter</groupId>
+                            <artifactId>junit-jupiter-engine</artifactId>
+                            <version>5.0.2</version>
```
Comment (chibenwa) [inline pom.xml]: I would introduce version variable as it is repeated in surefire plugin: we might forget to update this later on. A variable will protect us from this.

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+            .of(new MailAddress("rec1@domain.com"),
+                new MailAddress("rec2@domain.com"));
+        MailAddress sender = new MailAddress("sender@domain.com");
+        Mail mail = new MailImpl(name, sender, recipients, mailContent);
+        mail.setAttribute(TEST_ATTRIBUTE, "testValue");
+        return mail;
+    }
+
+
+    default InputStream generateMailContent(String body) {
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java]: Hand crafting a MimeMessage bytes looks stange, low level and obfuscated. I would rather go for a MimeMessage, using the **MimeMessageBuilder**.

MailImpl has a constructor that takes a MimeMessage, with the very same other parameters. That seems to be the way to go.

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    @Test
+    default void retrievingUnknownMailShouldReturnNull() throws Exception {
+        MailRepository testee = retrieveRepository();
+        assertThat(testee.retrieve("random")).isNull();
+    }
+
+    @Test
+    default void retrievingUnknownMailShouldHaveNoEffect() throws Exception {
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java]: removingUnknownMailShouldHaveNoEffect

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        MailRepository testee = retrieveRepository();
+        testee.remove("random");
+    }
+
+    @Test
+    default void listShouldReturnStoredMailsKeys() throws Exception {
+        MailRepository testee = retrieveRepository();
+        testee.store(createMail("mail1"));
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java]: Extract in variable?

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        testee.store(createMail("mail2"));
+        assertThat(testee.list()).containsExactly("mail1", "mail2");
+    }
+
+    @Test
+    default void storingMessageWithSameKeyTwiceShouldUpdateMessageContent() throws Exception {
+        MailRepository testee = retrieveRepository();
+        testee.store(createMail("mail1"));
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java]: idem?

**Comment by chibenwa** [comment]
I still do have some comments on this changeset. I will contribute the fixes.

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java (hunk 7 above)
Comment (mbaechler) [reply server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java]: good move.

File: server/data/data-jdbc/pom.xml [hunk 11]
```diff
@@ -37,6 +37,12 @@
             <groupId>${project.groupId}</groupId>
             <artifactId>apache-mailet-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>apache-mailet-base</artifactId>
```
Comment (mbaechler) [inline server/data/data-jdbc/pom.xml]: I don't like this dependency, it's very close to a circular dependency. Couldn't we move this MimeMessageBuilder elsewhere instead ?

File: server/mailrepository/mailrepository-api/pom.xml [hunk 12]
```diff
@@ -37,6 +37,12 @@
             <groupId>${project.groupId}</groupId>
             <artifactId>apache-mailet-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>apache-mailet-base</artifactId>
```
Comment (mbaechler) [inline server/mailrepository/mailrepository-api/pom.xml]: same thing here

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java [hunk 13]
```diff
@@ -79,33 +79,39 @@ default void checkMailEquality(Mail actual, Mail expected) {
...
         Mail email = createMail("mail1");
+
         testee.store(email);
     }
 
     @Test
     default void retrieveShouldGetStoredMail() throws Exception {
         MailRepository testee = retrieveRepository();
         String key1 = "mail1";
+
```
Comment (mbaechler) [inline server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java]: this one is misplaced, should be before `store` call

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java [hunk 14]
```diff
@@ -114,8 +120,10 @@ default void listShouldReturnStoredMailsKeys() throws Exception {
         MailRepository testee = retrieveRepository();
         String key1 = "mail1";
         String key2 = "mail2";
+
         testee.store(createMail(key1));
         testee.store(createMail(key2));
```
Comment (mbaechler) [inline server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java]: only the second call should be in that black

**Comment by mbaechler** [comment]
one more round of changes, I mainly moved the MimeMessageBuilder to core. (github order is wrong)

**Comment by chibenwa** [comment]
test this please (Jenkins history discarded)

File: server/data/data-jdbc/src/test/java/org/apache/james/mailrepository/jdbc/JDBCMailRepositoryTest.java [hunk 15]
```diff
@@ -60,4 +75,9 @@ private BasicDataSource getDataSource() {
         return ds;
     }
 
+    @Test
+    @Disabled("JDBC doesn't update the message Content")
```
Comment (chibenwa) [inline server/data/data-jdbc/src/test/java/org/apache/james/mailrepository/jdbc/JDBCMailRepositoryTest.java]: Where is the ticket for this?

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+            () -> assertThat(actual.getAttribute(TEST_ATTRIBUTE)).isEqualTo(expected.getAttribute(TEST_ATTRIBUTE))
+        );
+    }
+
+    MailRepository retrieveRepository() throws Exception;
+
+    @Test
+    default void storeRegularMailShouldNotFail() throws Exception {
+        MailRepository testee = retrieveRepository();
+        Mail email = createMail("mail1");
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java]: I thought you did not wanted me to use **email** but rather **mail** ;-)

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+
+        testee.store(email);
+    }
+
+    @Test
+    default void retrieveShouldGetStoredMail() throws Exception {
+        MailRepository testee = retrieveRepository();
+        String key1 = "mail1";
+        Mail email = createMail(key1);
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java]: mail

File: server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+    @Test
+    default void listShouldReturnStoredMailsKeys() throws Exception {
+        MailRepository testee = retrieveRepository();
+        String key1 = "mail1";
+        String key2 = "mail2";
+        testee.store(createMail(key1));
+
+        testee.store(createMail(key2));
+
+        assertThat(testee.list()).containsExactly(key1, key2);
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java]: Isn't exactly a bit tough?

(How do you define order?)

Please relax that assertion.

**Pull request merged**